gimprc.5.in
.\" Hey Emacs! This file is -*- nroff -*- source.
.\"
.\" This manpage is Copyright (C) 1999 Austin Donnelly <austin@gimp.org>
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the gimp is constantly changing, this manual page may be
.\" incorrect or out-of-date.  The author(s) assume no responsibility
.\" for errors or omissions, or for damages resulting from the use of
.\" the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Sun Feb  7 22:22:38 GMT 1999 Austin Donnelly <austin@gimp.org>
.\"   * initial revision
.\"
.\" Sat Feb 13 23:47:36 GMT 1999 Austin Donnelly <austin@gimp.org>
.\"   * added section on path expansion, and renamed to gimprc.5.in
.\"
.\" Wed Mar  3 13:09:52 MET 1999 Sven Neumann <sven@gimp.org>
.\"   * reflect the changes that were made to the resolution stuff
.\"
.\" Thu Oct 12 20:14:36 PDT 2000 Ben Gertzfield <che@debian.org>
.\"   * Move gimprc to gimpsysconfdir, make FILES section reflect this
.\"   * Removed .gimp-1.1 hard ref, replace with gimpdir
.\"
.TH GIMPRC 5 "14 Oct 2000" "Version @GIMP_VERSION@" "GIMP Manual Pages"
.SH NAME
gimprc \- gimp configuration file
.SH DESCRIPTION
The
.B gimprc
file is a configuation file read by the gimp when it starts up.  There
are two of these: one system-wide one stored in
@gimpsysconfdir@/gimprc and a per-user \fB$HOME\fP/@gimpdir@/gimprc
which may override system settings.

Comments are introduced by a hash sign (#), and continue until the
end of the line.  Blank lines are ignored.

The
.B gimprc
file associates values with properties.  These properties may be set
by lisp-like assignments of the form:
.IP
\f3(\f2property-name\ \ \ value\f3)\f1
.TP
where:
.TP 10
.I property-name
is one of the property names described below.
.TP
.I value
is the value the property is to be set to.
.PP

Either spaces or tabs may be used to separate the name from the value.

The values have an associated type, described below:
.TP
.I STRING
A sequence of characters surrounded by double-quotes (").  A backslash
(\\) may be used to escape either double-quote or itself to generate a
literal double-quote or a literal backslash.
.TP
.I PATH
Same as STRING, but path-expansion (see below) is also performed.
.TP
.I DOUBLE
An optional minus sign (-) followed by zero or more decimal digits,
and optionally a decimal-point (.) followed by zero or more decimal
digits.
.TP
.I FLOAT
Same as DOUBLE.
.TP
.I INT
Same as DOUBLE, except the value is rounded to the next lowest integer.
.TP
.I BOOLEAN
One of
.BR true ", " on ", " yes ", "
.BR false ", " off ", or " no "."
Alternatively, the empty string is interpreted to mean
.BR true "."
.TP
.I POSITION
Same as two INTs separated by whitespace.
.TP
.I MEMSIZE
An INT followed by a size specifier.  A size specifier is one of
.BR m ", " M ", "
.BR k ", " K ", "
.BR b ", or " B "."
The size specified may be omitted, in which case it defaults to
.BR k "."
.TP
.I IMAGETYPE
One of
.BR rgb ", " grey ", or " gray "."
.TP
.I COLORCUBE
Four whitespace separated INTs, giving the number of shades of red,
green, blue and grey (respectively) in the color cube.
.TP
.I PREVIEWSIZE
One of
.BR none ", "tiny ", " small ", " medium ", "large " or " huge ", "
or an INT.  tiny is 24x24, small is 32x32, medium is 48x48, 
large is 64x64 and huge is 128x128.
.TP
.I UNIT
One of
.BR inches ", " millimeters ", " picas ", or " points
or the identifier of one of the units you have defined in your user
units database.
.PP
.SH PROPERTIES
Valid properties and their types:
.TP
.I temp-path PATH
Set the temporary storage directory. Files will appear here
during the course of running the gimp.  Most files will disappear
when the gimp exits, but some files are likely to remain,
such as working palette files, so it is best if this directory
not be one that is shared by other users or is cleared on machine
reboot such as /tmp.
.TP
.I swap-path PATH
Set the swap file location. The gimp uses a tile based memory
allocation scheme. The swap file is used to quickly and easily
swap tiles out to disk and back in. Be aware that the swap file
can easily get very large if the gimp is used with large images.
Also, things can get horribly slow if the swap file is created on 
a directory that is mounted over NFS.  For these reasons, it may 
be desirable to put your swap file in "/tmp".
.TP
.I brush-path PATH
Set the brush search path. This is a colon-separated list of
directories to be searched for brushes.
.TP
.I pattern-path PATH
Set the pattern search path.  This is a colon-separated list of
directories to be searched for patterns.
.TP
.I plug-in-path PATH
Set the plug-in search path. This is a colon-separated list of
directories which will be scanned at startup to register new plugins.
.TP
.I palette-path PATH
Set the palette search path.  This is a colon-separated list of
directories to be searched for palettes.
.TP
.I gradient-path PATH
Set the gradient search path.  This is a colon-separated list of
directories to be searched for gradients.
.TP
.I module-path PATH
Set the module search path. This is a colon-separated list of
directories which will be scanned at startup for modules to be loaded.
.TP
.I default-brush STRING
Specify a default brush.  This doesn't actually do anything any more,
since the default brush is set as part of the saved device status.
.TP
.I default-pattern STRING
Specify a default pattern. The pattern is searched for in the
specified pattern path.
.TP
.I default-palette STRING
Specify a default palette.  The palette is searched for in the
specified palette path.
.TP
.I default-gradient STRING
Specify a default gradient. The gradient is searched for in the
specified gradient path.
.TP
.I gamma-correction DOUBLE
Set the gamma correction value for the display.  1.0 corresponds to no
gamma correction.  For most displays, gamma correction should be set
to between 2.0 and 2.6 Run the utility "gamma_correct" to determine
appropriate values for your display.  XXX is this valid info?  One
important item to keep in mind: Many images that you might get from
outside sources will in all likelihood already be gamma-corrected.  In
these cases, the image will look washed-out if the gimp has
gamma-correction turned on.  If you are going to work with images of
this sort, turn gamma correction off by setting the value to 1.0.
.TP
.I color-cube COLORCUBE
Set the displays color cube.  No longer used in gimp-1.2.x since the
introduction of GdkRgb.
.TP
.I install-colormap BOOLEAN
Install a private colormap by default - not actually used anymore since
the introduction of GdkRgb.
.TP
.I tile-cache-size MEMSIZE
The tile cache is used to make sure the gimp doesn't thrash
tiles between memory and disk. Setting this value higher will
cause the gimp to use less swap space, but will also cause
the gimp to use more memory. Conversely, a smaller cache size
causes the gimp to use more swap space and less memory.
Note: the gimp will still run even if `tile-cache-size' is
set to 0. The actual size can contain a suffix of 'm', 'M',
'k', 'K', 'b' or 'B', which makes the gimp interpret the
size as being specified in megabytes, kilobytes and bytes
respectively. If no suffix is specified the size defaults to
being specified in kilobytes.
.TP
.I marching-ants-speed INT
Speed of marching ants in the selection outline.  This value is in
milliseconds (less time indicates faster marching).
.TP
.I last-opened-size INT
How many recently opened image filenames to keep on the File menu.
.TP
.I undo-levels INT
Set the number of operations kept on the undo stack.
.TP
.I transparency-type INT
Set the manner in which transparency is displayed in images.
Transparency type can be one of 0 - Light Checks, 1 - Mid-Tone Checks,
2 - Dark Checks, 3 - White Only, 4 - Gray Only, or 5 - Black Only.
.TP
.I transparency-size INT
Check size can be one of 0 - Small, 1 - Medium, or 2 - Large
.TP
.I perfect-mouse BOOLEAN
If set to true, the X server is queried for the mouse's current
position on each motion event, rather than relying on the position
hint.  This means painting with large brushes should be more accurate,
but it may be slower.  Perversely, on some X servers turning on this
option results in faster painting.
.TP
.I colormap-cycling BOOLEAN
Specify that marching ants for selected regions will be drawn with
colormap cycling as oposed to redrawing with different stipple masks.
This color cycling option works only with 8-bit displays.
.TP
.I default-threshold INT
Tools such as fuzzy-select and bucket fill find regions based on a
seed-fill algorithm.  The seed fill starts at the intially selected
pixel and progresses in all directions until the difference of pixel
intensity from the original is greater than a specified threshold.
This value represents the default threshold.
.TP
.I stingy-memory-use BOOLEAN
There is always a tradeoff between memory usage and speed.  In most
cases, the GIMP opts for speed over memory.  However, if memory is a
big issue, set stingy-memory-use.
.TP
.I allow-resize-windows BOOLEAN
When zooming into and out of images, this option enables the automatic
resizing of windows.
.TP
.I dont-allow-resize-windows BOOLEAN
Negated version of allow-resize-windows.
.TP
.I cursor-updating BOOLEAN
Context-dependent cursors are cool.  They are enabled by default.
However, they require overhead that you may want to do without.
.TP
.I no-cursor-updating BOOLEAN
Negated version of cursor-updating.
.TP
.I preview-size PREVIEWSIZE
Set the layer preview size.
.TP
.I show-rulers BOOLEAN
Set the ruler visibility. The default behavior is for rulers to be on.
This can also be toggled with the View->Toggle Rulers command or
shift+control+r.
.TP
.I dont-show-rulers BOOLEAN
Negated version of show-rulers.
.TP
.I show-statusbar BOOLEAN
Controlling statusbar visibility. The default behavior is to show the
statusbar.  This can also be toggled with the View->Toggle Statusbar
command or shift+control+s.
.TP
.I dont-show-statusbar BOOLEAN
Negated version of show-statusbar.
.TP
.I auto-save BOOLEAN
Auto saving is not yet implemented! Nothing will be auto-saved, no
matter how you set this.
.TP
.I dont-auto-save BOOLEAN
Negated version of auto-save.
.TP
.I cubic-interpolation BOOLEAN
Set the level of interpolation.  If set, this option enables cubic
interpolation when scaling or transforming.  By default, GIMP uses
linear interpolation, which is faster, but has poorer quality.
.TP
.I confirm-on-close BOOLEAN
Ask for confirmation before closing an image without saving. This is
the default.
.TP
.I dont-confirm-on-close BOOLEAN
Negated version of confirm-on-close.
.TP
.I save-session-info BOOLEAN
Remember the positions and sizes of the main dialogs and asks your
window-manager to place them there again the next time you use the
GIMP.
.TP
.I dont-save-session-info BOOLEAN
Negated version of save-session-info.
.TP
.I save-device-status BOOLEAN
Remember the current tool, pattern, color, and brush across GIMP
sessions.
.TP
.I dont-save-device-status BOOLEAN
Negated version of save-device-status.
.TP
.I always-restore-session BOOLEAN
Let GIMP try to restore your last saved session.
.TP
.I show-tips BOOLEAN
Set to display a handy GIMP tip on startup.
.TP
.I dont-show-tips BOOLEAN
Negated version of show-tips.
.TP
.I show-tool-tips BOOLEAN
Set to display tooltips in the toolbox.
.TP
.I dont-show-tool-tips BOOLEAN
Negated version of show-tool-tips.
.TP
.I default-image-size POSITION
Set the default image size in the File/New dialog.
.TP
.I default-image-type IMAGETYPE
Set the default image type in the File/New dialog.
.TP
.I default-units UNIT
Set the default units for new images and for the File/New dialog.
This units will be used for coordinate display when not in dot-for-dot
mode. The default units can be one of inches , millimeters, points or
picas or the identifier of one of the units defined in your user units
database. The default is inches.
.TP
.I default-xresolution FLOAT
Set the default horizontal resolution for new images and for the
File/New dialog. This value is always in dpi (dots per inch).
.TP
.I default-yresolution FLOAT
Set the default vertical resolution for new images and for the
File/New dialog. This value is always in dpi (dots per inch).
.TP
.I default-resolution-units UNIT
Set the units for the display of the default resolution in the
File/New dialog. The default resolution units can be one of inches,
millimeters, points or picas or the identifier of one of the units
defined in your user unit database. The default is inches.
.TP
.I monitor-xresolution FLOAT
Set the monitor's horizontal resolution, in dots per inch.  If set to
0, forces the X server to be queried for both horizontal and vertical
resolution information.
.TP
.I monitor-yresolution FLOAT
Set the monitor's vertical resolution, in dots per inch.  If set to
0, forces the X server to be queried for both horizontal and vertical
resolution information.
.TP
.I num-processors INT
On multiprocessor machines, if GIMP has been compiled with --enable-mp
this sets how many processors GIMP should use simultaneously.
.TP
.I image-title-format STRING
Set the text to appear in image window titles.  Certain % character
sequences are recognised and expanded as follows:

%%  literal percent sign
.br
%f  bare filename, or "Untitled"
.br
%F  full path to file, or "Untitled"
.br
%p  PDB image id
.br
%i  view instance number
.br
%t  image type (RGB, indexed, greyscale)
.br
%z  zoom factor as a percentage
.br
%s  source scale factor
.br
%d  destination scale factor
.br
%Dx expands to x if the image is dirty, the empty string
otherwise.  The x can be any character, eg %D*.

The default format string is "%f-%p.%i (%t)".

.PP
.SH PATH EXPANSION
Strings of type PATH are expanded in a manner similar to
.BR bash (1).
Specifically: tilde (~) is expanded to the user's home directory, and
${variable} is expanded to the current value of the variable.  Note
that the bash feature of being able to refer to other user's home
directories by writing ~userid/ is not valid in this file.

The only variable initially defined is
.I gimp_dir
, which is set to either the interned value
.B @gimpdir@
or the environment variable GIMP_DIRECTORY.  If the path in
GIMP_DIRECTORY is relative, it is considered relative to your home
directory.  The same variable expansion syntax can be used to refer to
environment variables.  New variables may be defined so long as their
name does not shadow one of the property names given in the previous
section.  Variables are set using the following syntax:

NEWS
                   ------------------------------
                   GNU Image Manipulation Program
                         Development Branch
                   ------------------------------

This is the development branch of GIMP.

Overview of Changes from GIMP 3.2.4 to GIMP 3.3.2
=================================================

Core:

  * Using native file choosers in various parts of the code.
  * Rendering MyPaint brush previews at best sizes to be non-blurry.
  * GimpViewable preview API now passes around the scale factor to allow
    best decision when rendering previews.
  * Error console save dialog now has a ".log" extension filter.
  * Improve list of "Recent Images" in the Welcome dialog, with no
    arbirtrary limitation of 8 rows.
  * More GtkTreeView replaced in preparation of GTK4 port.
  * Various preliminatory changes in preparation of GTK4 port.
  * macOS: touchpads are now less sensitive in pan gestures and behave as
    expected.
  * MSIX: it is now possible to open multiple images from the OS.
  * Stop noninteractive filters on invisible layers unless the relevant
    option is enabled in Preferences.
  * Various layer modes are now being ported to work on non-RGBA
    contents when it makes sense.
  * Non-GUI filters will be applied non-destructively on group layers as
    well as non-raster layers. For now they are still applied
    destructively only on raster layers, until a proper design decision
    is taken on this topic.
  * "Layers to Image Size" action is now only sensitive on raster
    layers.
  * Layer Masks can now have non-destructive filters (listed separately
    in the NDE popover).
  * Added "Rotation" input for dynamics. This gives us support of the
    Wacom Art Pen 2 stylus at least.
  * We now verify if a plug-in has been updated since last start by
    using the changed time (ctime) as fallback when the modified time
    (mtime) is 0. This is because on some environment (such as ostree,
    hence flatpak), all files mtimes are set to the UNIX epoch.
  * Added support of dashboard backtraces on Windows ARM64.
  * Krita KPL palette import support added.
  * Work started on a new project format.
  * Do not perform the (wrong) optimization replacing pass-through by
    the Normal layer mode for layers with effects.
  * Dashboard backtraces now support macOS.
  * Clear does not affect non-raster layers, unless rasterized.
  * New project format for GIMP (meant to replace XCF) is under
    development. Sub-Features:
    - More robust image recovery (e.g. after a crash, external energy
      loss or other issues) reimplemented based on the new format.
    - Constant, reliable and seamless auto-save of opened images.
  * About dialog: more info about the OS, dependencies and hardware is
    being shown. A "Copy" button allows to easily add the
    markdown-formatted info to clipboard for pasting in reports.
  * Improve internal logics to determine if an item is "writable"
    regarding actions modifying item contents.
  * "Reshow Filter" and "Reapply Filter" now work on group layers too.

Tools:

  * Gradient tool: new option "Editable gradient" to apply the gradient
    non-destructively.
  * Heal tool: healing behaves more as expected near edges.
  * Text tool:
    - Allow selecting with Shift-Click.
    - Preview text layer content while Alt-dragging.
    - Ignore selected layer mask of a text layer when selecting it with
      the tool to edit the text contents. The layer mask still stays the
      selected item in such a case.
  * Transform tools: add a tooltip on the chain icon to better explain
    what it does.
  * MyPaint Brush tool: new "Spectral Blending" checkbox and "Pigment"
    slider to blend spectrally while painting.
  * Color Picker tool: there was an exception ignoring Sample Merged
    option when a single layer was visible (allowing to retrieve the
    exact palette color from indexed images). This exception has been
    further restricted to the absence of filters. Sample Merged, even of
    single layer images, should return the color after filters are
    applied.
  * Path tool: new "Auto-connect on click" checkbox setting (ON by
    default) allowing to decide whether to auto-connect, or not, when
    clicking on the end node (instead allowing users to readjust it).

Graphical User Interface:

  * Small reorganization in GimpStrokeEditor.
  * Welcome dialog: icons and links realigned on Welcome tab; widgets
    realigned in Contribute tab.
  * New "branding" colors added in AppStream metadata.
  * macOS: title bar follows scheme. This is true for native file
    dialogs on macOS as well.
  * Levels tool now listens to the icon size setting.
  * Roadmap menu item removed from Help > GIMP Online menu.
  * We do not rotate the checkerboard anymore when rotating the canvas,
    but also scrolling or zooming. This significantly improves performance
    when interacting with a rotated canvas.
  * GimpColorScale moved from hardcoded (bad!) to proper theme-based CSS
    styling.
  * Append "(read only)" to system resource (brushes, gradients,
    palettes…) titles when trying to edit them in their respective
    Editor dockables.
  * Action Search dialog:
    - The shortcuts hints are moved to the right side, similar to the
      rest of the application (e.g. shortcuts hints in menus).
    - A ".search-results" class was added for the results row, and theme
      was updated to highlight the hovered action row.
  * Icon picker widget: the file filter name is now titled "Image
    files".
  * Multiple selected layers are now added to a new group.
  * Navigation dockable: grab back focus on the view, after using the
    zoom slider.
  * Patterns which are smaller than the preview area (in 1 or 2
    dimensions) will now fill the full area by repeating themselves,
    since they are meant to be tileable anyway. It makes for nicer
    pattern previews.
  * Size entry in Scale and Print Size dialogs receive the focus by

app/config/gimprc.c
/* GIMP - The GNU Image Manipulation Program
 * Copyright (C) 1995 Spencer Kimball and Peter Mattis
 *
 * GimpRc, the object for GIMPs user configuration file gimprc.
 * Copyright (C) 2001-2002  Sven Neumann <sven@gimp.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <string.h>

#include <gio/gio.h>
#include <gegl.h>

#include "libgimpbase/gimpbase.h"
#include "libgimpconfig/gimpconfig.h"

#include "config-types.h"

#include "gimpconfig-file.h"
#include "gimprc.h"
#include "gimprc-deserialize.h"
#include "gimprc-serialize.h"
#include "gimprc-unknown.h"

#include "gimp-intl.h"


enum
{
  PROP_0,
  PROP_VERBOSE,
  PROP_SYSTEM_GIMPRC,
  PROP_USER_GIMPRC
};


static void         gimp_rc_config_iface_init (GimpConfigInterface *iface);

static void         gimp_rc_dispose           (GObject      *object);
static void         gimp_rc_finalize          (GObject      *object);
static void         gimp_rc_set_property      (GObject      *object,
                                               guint         property_id,
                                               const GValue *value,
                                               GParamSpec   *pspec);
static void         gimp_rc_get_property      (GObject      *object,
                                               guint         property_id,
                                               GValue       *value,
                                               GParamSpec   *pspec);

static GimpConfig * gimp_rc_duplicate         (GimpConfig   *object);
static gboolean     gimp_rc_idle_save         (GimpRc       *rc);
static void         gimp_rc_notify            (GimpRc       *rc,
                                               GParamSpec   *param,
                                               gpointer      data);


G_DEFINE_TYPE_WITH_CODE (GimpRc, gimp_rc, GIMP_TYPE_PLUGIN_CONFIG,
                         G_IMPLEMENT_INTERFACE (GIMP_TYPE_CONFIG,
                                                gimp_rc_config_iface_init))

#define parent_class gimp_rc_parent_class


static void
gimp_rc_class_init (GimpRcClass *klass)
{
  GObjectClass *object_class = G_OBJECT_CLASS (klass);

  object_class->dispose      = gimp_rc_dispose;
  object_class->finalize     = gimp_rc_finalize;
  object_class->set_property = gimp_rc_set_property;
  object_class->get_property = gimp_rc_get_property;

  g_object_class_install_property (object_class, PROP_VERBOSE,
                                   g_param_spec_boolean ("verbose",
                                                         NULL, NULL,
                                                         FALSE,
                                                         G_PARAM_READWRITE |
                                                         G_PARAM_CONSTRUCT));

  g_object_class_install_property (object_class, PROP_SYSTEM_GIMPRC,
                                   g_param_spec_object ("system-gimprc",
                                                        NULL, NULL,
                                                        G_TYPE_FILE,
                                                        G_PARAM_READWRITE |
                                                        G_PARAM_CONSTRUCT));

  g_object_class_install_property (object_class, PROP_USER_GIMPRC,
                                   g_param_spec_object ("user-gimprc",
                                                        NULL, NULL,
                                                        G_TYPE_FILE,
                                                        G_PARAM_READWRITE |
                                                        G_PARAM_CONSTRUCT));
}

static void
gimp_rc_init (GimpRc *rc)
{
  rc->autosave     = FALSE;
  rc->save_idle_id = 0;
}

static void
gimp_rc_dispose (GObject *object)
{
  GimpRc *rc = GIMP_RC (object);

  if (rc->save_idle_id)
    gimp_rc_idle_save (rc);

  G_OBJECT_CLASS (parent_class)->dispose (object);
}

static void
gimp_rc_finalize (GObject *object)
{
  GimpRc *rc = GIMP_RC (object);

  g_clear_object (&rc->system_gimprc);
  g_clear_object (&rc->user_gimprc);

  G_OBJECT_CLASS (parent_class)->finalize (object);
}

static void
gimp_rc_set_property (GObject      *object,
                      guint         property_id,
                      const GValue *value,
                      GParamSpec   *pspec)
{
  GimpRc *rc = GIMP_RC (object);

  switch (property_id)
    {
    case PROP_VERBOSE:
      rc->verbose = g_value_get_boolean (value);
      break;

    case PROP_SYSTEM_GIMPRC:
      g_set_object (&rc->system_gimprc, g_value_get_object (value));
      if (! rc->system_gimprc)
        rc->system_gimprc = gimp_sysconf_directory_file ("gimprc", NULL);
      break;

    case PROP_USER_GIMPRC:
      g_set_object (&rc->user_gimprc, g_value_get_object (value));
      if (! rc->user_gimprc)
        rc->user_gimprc = gimp_directory_file ("gimprc", NULL);
      break;

   default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
    }
}

static void
gimp_rc_get_property (GObject    *object,
                      guint       property_id,
                      GValue     *value,
                      GParamSpec *pspec)
{
  GimpRc *rc = GIMP_RC (object);

  switch (property_id)
    {
    case PROP_VERBOSE:
      g_value_set_boolean (value, rc->verbose);
      break;
    case PROP_SYSTEM_GIMPRC:
      g_value_set_object (value, rc->system_gimprc);
      break;
    case PROP_USER_GIMPRC:
      g_value_set_object (value, rc->user_gimprc);
      break;

   default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
    }
}

static void
gimp_rc_config_iface_init (GimpConfigInterface *iface)
{
  iface->serialize   = gimp_rc_serialize;
  iface->deserialize = gimp_rc_deserialize;
  iface->duplicate   = gimp_rc_duplicate;
}

static void
gimp_rc_duplicate_unknown_token (const gchar *key,
                                 const gchar *value,
                                 gpointer     user_data)
{
  gimp_rc_add_unknown_token (GIMP_CONFIG (user_data), key, value);
}

static GimpConfig *
gimp_rc_duplicate (GimpConfig *config)
{
  GObject    *gimp;
  GimpConfig *dup;

  g_object_get (config, "gimp", &gimp, NULL);

  dup = g_object_new (GIMP_TYPE_RC,
                      "gimp", gimp,
                      NULL);

  if (gimp)
    g_object_unref (gimp);

  gimp_config_sync (G_OBJECT (config), G_OBJECT (dup), 0);

  gimp_rc_foreach_unknown_token (config,
                                 gimp_rc_duplicate_unknown_token, dup);

  return dup;
}

static gboolean
gimp_rc_idle_save (GimpRc *rc)
{
  gimp_rc_save (rc);

  rc->save_idle_id = 0;

  return FALSE;
}

static void
gimp_rc_notify (GimpRc     *rc,
                GParamSpec *param,
                gpointer    data)
{
  if (!rc->autosave)
    return;

  if (!rc->save_idle_id)
    rc->save_idle_id = g_idle_add ((GSourceFunc) gimp_rc_idle_save, rc);
}

/**
 * gimp_rc_new:
 * @gimp:          a #Gimp
 * @system_gimprc: the name of the system-wide gimprc file or %NULL to
 *                 use the standard location
 * @user_gimprc:   the name of the user gimprc file or %NULL to use the
 *                 standard location
 * @verbose:       enable console messages about loading and saving
 *
 * Creates a new GimpRc object and loads the system-wide and the user
 * configuration files.
 *
 * Returns: the new #GimpRc.
 */
GimpRc *
gimp_rc_new (GObject  *gimp,
             GFile    *system_gimprc,
             GFile    *user_gimprc,
             gboolean  verbose)
{
  GimpRc *rc;

  g_return_val_if_fail (G_IS_OBJECT (gimp), NULL);
  g_return_val_if_fail (system_gimprc == NULL || G_IS_FILE (system_gimprc),
                        NULL);
  g_return_val_if_fail (user_gimprc == NULL || G_IS_FILE (user_gimprc),
                        NULL);

  rc = g_object_new (GIMP_TYPE_RC,
                     "gimp",          gimp,
                     "verbose",       verbose,
                     "system-gimprc", system_gimprc,
                     "user-gimprc",   user_gimprc,
                     NULL);

  gimp_rc_load_system (rc);
  gimp_rc_load_user (rc);

  return rc;
}

void
gimp_rc_load_system (GimpRc *rc)
{
  GError *error = NULL;

  g_return_if_fail (GIMP_IS_RC (rc));

  if (rc->verbose)
    g_print ("Parsing '%s'\n",
             gimp_file_get_utf8_name (rc->system_gimprc));

  if (! gimp_config_deserialize_file (GIMP_CONFIG (rc),
                                      rc->system_gimprc, NULL, &error))
    {
      if (error->code != GIMP_CONFIG_ERROR_OPEN_ENOENT)
        g_message ("%s", error->message);

      g_clear_error (&error);
    }
}

void
gimp_rc_load_user (GimpRc *rc)
{
  GError *error = NULL;

  g_return_if_fail (GIMP_IS_RC (rc));

  if (rc->verbose)
    g_print ("Parsing '%s'\n",
             gimp_file_get_utf8_name (rc->user_gimprc));

  if (! gimp_config_deserialize_file (GIMP_CONFIG (rc),
                                      rc->user_gimprc, NULL, &error))
    {
      if (error->code != GIMP_CONFIG_ERROR_OPEN_ENOENT)
        {
          g_message ("%s", error->message);

          gimp_config_file_backup_on_error (rc->user_gimprc, "gimprc", NULL);
        }

      g_clear_error (&error);
    }
}

void
gimp_rc_set_autosave (GimpRc   *rc,
                      gboolean  autosave)
{
  g_return_if_fail (GIMP_IS_RC (rc));

  autosave = autosave ? TRUE : FALSE;

  if (rc->autosave == autosave)
    return;

  if (autosave)
    g_signal_connect (rc, "notify",
                      G_CALLBACK (gimp_rc_notify),
                      NULL);
  else
    g_signal_handlers_disconnect_by_func (rc, gimp_rc_notify, NULL);

  rc->autosave = autosave;
}


/**
 * gimp_rc_query:
 * @rc:  a #GimpRc object.
 * @key: a string used as a key for the lookup.
 *
 * This function looks up @key in the object properties of @rc. If
 * there's a matching property, a string representation of its value
 * is returned. If no property is found, the list of unknown tokens
 * attached to the @rc object is searched.
 *
 * Returns: (nullable): a newly allocated string representing the value or %NULL
 *               if the key couldn't be found.
 **/
gchar *
gimp_rc_query (GimpRc      *rc,
               const gchar *key)
{
  GObjectClass  *klass;
  GObject       *rc_object;
  GParamSpec   **property_specs;
  GParamSpec    *prop_spec;
  guint          i, n_property_specs;
  gchar         *retval = NULL;

  g_return_val_if_fail (GIMP_IS_RC (rc), NULL);
  g_return_val_if_fail (key != NULL, NULL);

  rc_object = G_OBJECT (rc);
  klass = G_OBJECT_GET_CLASS (rc);

  property_specs = g_object_class_list_properties (klass, &n_property_specs);

  if (!property_specs)
    return NULL;

  for (i = 0, prop_spec = NULL; i < n_property_specs && !prop_spec; i++)
    {
      prop_spec = property_specs[i];

      if (! (prop_spec->flags & GIMP_CONFIG_PARAM_SERIALIZE) ||
          strcmp (prop_spec->name, key))
        {
          prop_spec = NULL;
        }
    }

  if (prop_spec)
    {
      GString *str   = g_string_new (NULL);
      GValue   value = G_VALUE_INIT;

      g_value_init (&value, prop_spec->value_type);
      g_object_get_property (rc_object, prop_spec->name, &value);

      if (gimp_config_serialize_value (&value, str, FALSE))
        retval = g_string_free (str, FALSE);
      else
        g_string_free (str, TRUE);

      g_value_unset (&value);
    }
  else
    {
      retval = g_strdup (gimp_rc_lookup_unknown_token (GIMP_CONFIG (rc), key));
    }

  g_free (property_specs);

  if (!retval)
    {
      const gchar * const path_tokens[] =
      {
        "gimp_dir",
        "gimp_data_dir",
        "gimp_plug_in_dir",
        "gimp_plugin_dir",
        "gimp_sysconf_dir"
      };

      for (i = 0; !retval && i < G_N_ELEMENTS (path_tokens); i++)
        if (strcmp (key, path_tokens[i]) == 0)
          retval = g_strdup_printf ("${%s}", path_tokens[i]);
    }

  if (retval)
    {
      gchar *tmp = gimp_config_path_expand (retval, FALSE, NULL);

      if (tmp)
        {
          g_free (retval);
          retval = tmp;
        }
    }

  return retval;
}

/**
 * gimp_rc_set_unknown_token:
 * @gimprc: a #GimpRc object.
 * @token:
 * @value:
 *
 * Calls gimp_rc_add_unknown_token() and triggers an idle-save if
 * autosave is enabled on @gimprc.
 **/
void
gimp_rc_set_unknown_token (GimpRc      *rc,
                           const gchar *token,
                           const gchar *value)
{
  g_return_if_fail (GIMP_IS_RC (rc));

  gimp_rc_add_unknown_token (GIMP_CONFIG (rc), token, value);

  if (rc->autosave)
    gimp_rc_notify (rc, NULL, NULL);
}

/**
 * gimp_rc_save:
 * @gimprc: a #GimpRc object.
 *
 * Saves any settings that differ from the system-wide defined
 * defaults to the users personal gimprc file.
 **/
void
gimp_rc_save (GimpRc *rc)
{
  GObject *gimp;
  GimpRc  *global;
  gchar   *header;
  GError  *error = NULL;

  const gchar *top =
    "GIMP gimprc\n"
    "\n"
    "This is your personal gimprc file.  Any variable defined in this file "
    "takes precedence over the value defined in the system-wide gimprc: ";
  const gchar *bottom =
    "\n"
    "Most values can be set within GIMP by changing some options in "
    "the Preferences dialog.";
  const gchar *footer =
    "end of gimprc";

  g_return_if_fail (GIMP_IS_RC (rc));

  g_object_get (rc, "gimp", &gimp, NULL);

  global = g_object_new (GIMP_TYPE_RC,
                         "gimp", gimp,
                         NULL);

  if (gimp)
    g_object_unref (gimp);

  gimp_config_deserialize_file (GIMP_CONFIG (global),
                                rc->system_gimprc, NULL, NULL);

  header = g_strconcat (top, gimp_file_get_utf8_name (rc->system_gimprc),
                        bottom, NULL);

  if (rc->verbose)
    g_print ("Writing '%s'\n",
             gimp_file_get_utf8_name (rc->user_gimprc));

  if (! gimp_config_serialize_to_file (GIMP_CONFIG (rc),
                                       rc->user_gimprc,
                                       header, footer, global,
                                       &error))
    {
      g_message ("%s", error->message);
      g_error_free (error);
    }

  g_free (header);
  g_object_unref (global);
}
